tests/e2e/tests/pilot/istio_rbac_test.go

```diff
@@ -56,55 +58,105 @@ func TestRBAC(t *testing.T) {
+		{dst: "a", src: "d", port: 90, allow: false},
+		{dst: "a", src: "d", port: 9090, allow: false},
+
+		{dst: "b", src: "a", path: "/xyz", allow: true},
+		{dst: "b", src: "a", path: "/", allow: true},
+		{dst: "b", src: "a", port: 90, allow: true},
+		{dst: "b", src: "a", port: 9090, allow: true},
+		{dst: "b", src: "c", path: "/", allow: true},
+		{dst: "b", src: "c", port: 90, allow: true},
+		{dst: "b", src: "c", port: 9090, allow: true},
+		{dst: "b", src: "d", path: "/", allow: true},
+		{dst: "b", src: "d", port: 90, allow: true},
+		{dst: "b", src: "d", port: 9090, allow: true},
+
+		{dst: "c", src: "a", path: "/", allow: false},
+		{dst: "c", src: "a", path: "/good", allow: false},
+		{dst: "c", src: "a", path: "/prefixXYZ", allow: false},
+		{dst: "c", src: "a", path: "/xyz/suffix", allow: false},
+		{dst: "c", src: "a", port: 90, allow: false},
+		{dst: "c", src: "a", port: 9090, allow: false},
+
+		{dst: "c", src: "b", path: "/", allow: false},
+		{dst: "c", src: "b", path: "/good", allow: false},
+		{dst: "c", src: "b", path: "/prefixXYZ", allow: false},
+		{dst: "c", src: "b", path: "/xyz/suffix", allow: false},
+		{dst: "c", src: "b", port: 90, allow: false},
+		{dst: "c", src: "b", port: 9090, allow: false},
+
+		{dst: "c", src: "d", path: "/", allow: false},
+		{dst: "c", src: "d", path: "/xyz", allow: false},
+		{dst: "c", src: "d", path: "/good", allow: allow},
+		{dst: "c", src: "d", path: "/prefixXYZ", allow: allow},
+		{dst: "c", src: "d", path: "/xyz/suffix", allow: allow},
+		{dst: "c", src: "d", port: 90, allow: allow},
+		{dst: "c", src: "d", port: 9090, allow: false},
+
+		{dst: "d", src: "a", path: "/xyz", allow: allow},
+		{dst: "d", src: "a", port: 90, allow: false},
+		{dst: "d", src: "a", port: 9090, allow: true},
+		{dst: "d", src: "b", path: "/", allow: allow},
+		{dst: "d", src: "b", port: 90, allow: false},
+		{dst: "d", src: "b", port: 9090, allow: true},
+		{dst: "d", src: "c", path: "/", allow: allow},
+		{dst: "d", src: "c", port: 90, allow: false},
+		{dst: "d", src: "c", port: 9090, allow: true},
 	}
 
 	for _, req := range cases {
 		for cluster := range tc.Kube.Clusters {
-			testName := fmt.Sprintf("%s from %s cluster->%s%s[%s]", req.src, cluster, req.dst, req.path, req.expect)
-			runRetriableTest(t, cluster, testName, defaultRetryBudget, func() error {
-				resp := ClientRequest(cluster, req.src, fmt.Sprintf("http://%s%s", req.dst, req.path), 1, "")
-				if len(resp.Code) > 0 && resp.Code[0] == req.expect {
-					return nil
+			port := ""
+			if req.port != 0 {
+				port = fmt.Sprintf(":%d", req.port)
+			}
+			expectStr := "deny"
+			if req.allow {
+				expectStr = "allow"
+			}
+			testName := fmt.Sprintf("%s from %s cluster->%s%s%s[%s]",
+				req.src, cluster, req.dst, req.path, port, expectStr)
+
+			runRetriableTest(t, cluster, testName, 30, func() error {
+				reqPath := fmt.Sprintf("http://%s%s%s", req.dst, port, req.path)
+				if !req.allow && port != "" {
```

Comment:
port != "" means TCP service? Can the test case specify both "port" and "path"?

Another question is, do we need to check the success case for TCP service?

Reply:
Yes, `port != ""` means the request will go through the tcp rbac filter. We can specify both port and path, but that doesn't make too much sense as we cannot enforce any access control on the path when it's going through the tcp rbac filter. Actually when path is "", I guess it's default to "/".

We do check the success case for TCP service, note the first if statement `!req.allow && port != ""` is true when the it's a tcp request and we're expecting a deny result.

The next else statement in line 150 will cover both success/fail http requests and success tcp requests. In that case, we can check the response code directly as the test client is behaving like a HTTP server even for those tcp ports.

Reply:
Sounds good. Thanks for explanation.